File: src/preview_controller.cpp
#include "preview_controller.h"
using namespace std;

PreviewController::PreviewController(double dt, double time_horizon)
    : dt_(dt), time_horizon_(time_horizon)
{
    NL_ = static_cast<int>(time_horizon_);
    
    // Initialize state and error integral
    state_.setZero();
    error_integral_.setZero();
}

PreviewController::~PreviewController()
{
}

void PreviewController::init(float vrp_height)
{
    // System matrices A, B, C
    A_ <<   1, dt_, pow(dt_, 2) / 2,
            0, 1, dt_,
            0, 0, 1;

    B_ <<   pow(dt_, 3) / 6,
            pow(dt_, 2) / 2,
            dt_;

    C_ << 1, 0, -vrp_height / 9.81;

    // Augmented system matrices (MATLAB: A_bar = [I_bar, F_bar])
    // A_bar (4x4): col 0 = [1;0;0;0], cols 1-3 = [C*A; A]
    Eigen::Vector4d I_bar;
    I_bar << 1, 0, 0, 0;

    Eigen::Matrix4d A_bar;
    A_bar.col(0) = I_bar;
    A_bar.block<1, 3>(0, 1) = C_ * A_;
    A_bar.block<3, 3>(1, 1) = A_;

    // B_bar (4x1): [C*B; B]
    Eigen::Vector4d B_bar;
    B_bar(0) = (C_ * B_)(0);
    B_bar.tail<3>() = B_;

    // Cost matrices: Q_bar has Q_e=1 at (0,0), R=1e-6
    Eigen::Matrix4d Q_bar = Eigen::Matrix4d::Zero();
    Q_bar(0, 0) = 1.0;
    const double R = 1e-6;

    // Solve DARE iteratively: K = A'KA - A'KB(R+B'KB)^{-1}B'KA + Q
    Eigen::Matrix4d K = Q_bar;
    for (int iter = 0; iter < 200000; ++iter)
    {
        double S_iter = R + (B_bar.transpose() * K * B_bar)(0, 0);
        Eigen::Matrix4d K_new = A_bar.transpose() * K * A_bar
                              - (A_bar.transpose() * K * B_bar) * (1.0 / S_iter) * (B_bar.transpose() * K * A_bar)
                              + Q_bar;
        double err = (K_new - K).norm();
        K = K_new;
        if (err < 1e-10)
            break;
    }
    K_dare_ = K;

    // Optimal gain: G = (R + B'KB)^{-1} * B' * K * A_bar  (1x4)
    double S = R + (B_bar.transpose() * K * B_bar)(0, 0);
    Eigen::RowVector4d G = (1.0 / S) * (B_bar.transpose() * K * A_bar);

    G_i_dare_ = G(0);
    G_x_dare_ = G.segment(1, 3);  // 1x3

    // Ac_bar_T = (A_bar - B_bar * G)'
    Eigen::Matrix4d Ac_bar_T = (A_bar - B_bar * G).transpose();

    // RBT = (R + B'KB)^{-1} * B'  (1x4)
    Eigen::RowVector4d RBT = (1.0 / S) * B_bar.transpose();

    // X_dare_[0] = -Ac_bar_T * K * I_bar
    X_dare_.setZero(NL_, 4);
    X_dare_.row(0) = (-Ac_bar_T * K * I_bar).transpose();

    G_d_dare_.setZero(NL_);
    G_d_dare_(0) = -G_i_dare_;

    for (int l = 1; l < NL_; ++l)
    {
        X_dare_.row(l) = (Ac_bar_T * X_dare_.row(l - 1).transpose()).transpose();
        G_d_dare_(l) = RBT * X_dare_.row(l - 1).transpose();
    }
}

Eigen::MatrixXd PreviewController::compute_target_state(const Eigen::MatrixXd &vrp_ref)
{
    Eigen::RowVector3d current_output = C_ * state_; 
    error_integral_ += (current_output - vrp_ref.row(0)).transpose();

    Eigen::RowVector3d term1 = -G_i_dare_ * error_integral_.transpose();
    Eigen::RowVector3d term2 = -G_x_dare_ * state_;
    Eigen::RowVector3d term3 = Eigen::RowVector3d::Zero();
    for (int i = 1; i < NL_; ++i)
        term3 -= G_d_dare_(i) * vrp_ref.row(i);

    Eigen::RowVector3d input = term1 + term2 + term3;
    Eigen::Matrix3d next_state = A_ * state_ + B_ * input;
    state_ = next_state;
    return state_;
}

void PreviewController::update_state(const Eigen::MatrixXd &next_state)
{
    state_ = next_state;
}

File: src/preview_controller.py
import torch

class preview_controller:
    def __init__(self, num_envs: int, device: torch.device, time_horizon: float = 2, dt: float = 0.008):
        self.num_envs = num_envs
        self.device = device
        self.time_horizon = time_horizon
        self.dt = dt
        self.hz = int(1/dt)
        self.NL = int(time_horizon*self.hz)

        X_dare = torch.zeros((self.NL,4,1), device=self.device)
        X_dare[0] = torch.tensor([  -69.0813009391229,
                                    -2420.65372019103,
                                    -669.387885399942,
                                    -2.72127899669319], device=self.device).reshape(4, 1)
        ## this parameter is tuned for com height: 0.728m
        self.G_d_dare = torch.zeros((self.NL,), device=self.device)

        self.K_dare = torch.tensor([[70.0813009391222,	2420.65372019101,	669.387885399937,	2.72127899669319],
                                    [2420.65372019101,	85969.0761591658,	23782.1340369413,	99.0855302680049],
                                    [669.387885399937,	23782.1340369413,	6579.12894080612,	27.4486400273521],
                                    [2.72127899669319,	99.0855302680049,	27.4486400273521,	0.125016968799296]],    
                                    device=self.device).unsqueeze(dim=0)
        
        self.G_i_dare = 556.382091536873
        self.G_x_dare = torch.tensor([[38991.9807941560,	11086.4028841705	,130.234568101964]], device=self.device).unsqueeze(dim=0)
        self.G_d_dare[0] = -self.G_i_dare

        Ac_bar_T = torch.tensor([[1.33026539679257,	-4.74779384778249e-05,	-0.0178042269291843,	-4.45105673229607],
                                 [24.1454286624088	,0.996672684305565,	-1.24774338541315	,-311.935846353287],
                                 [6.58882872047239,	0.00705396028721734,	0.645235107706503,	-88.6912230733744],
                                 [0.00312854461408964,	2.08866501886319e-05,	0.00383249382073698,	-0.0418765448157545]], 
                                device=self.device)
        RBT = torch.tensor([[-183.753752229875,	0.0264158747121758	,9.90595301706591	,2476.48825426648]], device=self.device)

        for l in torch.arange(start=1, end=self.NL):
            X_dare[l] = torch.mm(Ac_bar_T, X_dare[l-1])
            self.G_d_dare[l] = torch.mm(RBT, X_dare[l-1]).squeeze().squeeze()

        self.A = torch.tensor([[1, self.dt, pow(self.dt, 2)/2],
                               [0, 1, self.dt],
                               [0, 0, 1]], device=self.device).unsqueeze(dim=0)
        self.B = torch.tensor([[pow(self.dt,3)/6], [pow(self.dt, 2)/2], [self.dt]], device=self.device).unsqueeze(dim=0)
        self.C = torch.tensor([[1, 0, -0.728/9.81]], device=self.device).unsqueeze(dim=0)

        self.error_integral = torch.zeros((self.num_envs, 3), device=self.device)

        self.state = torch.zeros((self.num_envs, 3, 3), device=self.device)

    def compute_target_state(self, vrp_ref: torch.Tensor):
        with torch.no_grad():
            assert vrp_ref.shape == (3, self.num_envs, self.NL),  f"Given vrp ref shape : {vrp_ref.shape}"
            # 3, num_envs, NL -> NL, num_envs, 3
            vrp_ref_permute = vrp_ref.permute(2, 1, 0)

            self.error_integral += torch.bmm(self.C.repeat((self.num_envs, 1, 1)), self.state).squeeze() - vrp_ref_permute[0, :]

            input = -self.G_i_dare * self.error_integral \
                    -torch.bmm(self.G_x_dare.repeat((self.num_envs, 1, 1)), self.state).squeeze(dim=1) \
                    -(self.G_d_dare[1:self.NL].reshape(-1, 1, 1).repeat(1, self.num_envs, 3) * vrp_ref_permute[1:self.NL]).sum(dim=0)
            assert input.shape == (self.num_envs, 3), f"output shape is wrong, {input.shape}"
            
            return torch.bmm(self.A.repeat((self.num_envs, 1, 1)), self.state) + torch.bmm(self.B.repeat((self.num_envs, 1, 1)), input.unsqueeze(dim=1))

    def update_state(self, next_state: torch.Tensor):
        assert next_state.shape == (self.num_envs, 3, 3), f"next_state shape is wrong, {next_state.shape}"
        self.state = next_state.clone()